File: code/zato-web-admin/src/zato/admin/static/js/cloud/openstack/swift.js

// /////////////////////////////////////////////////////////////////////////////

$.fn.zato.data_table.OpenStackSwift = new Class({
    toString: function() {
        var s = '<OpenStackSwift id:{0} name:{1} is_active:{2} is_snet:{3}>';
        return String.format(s, this.id ? this.id : '(none)',
                                this.name ? this.name : '(none)',
                                this.is_active ? this.is_active : '(none)',
                                this.is_snet ? this.is_snet : '(none)'
                                );
    }
});

// /////////////////////////////////////////////////////////////////////////////

$(document).ready(function() {
    $('#data-table').tablesorter();
    $.fn.zato.data_table.class_ = $.fn.zato.data_table.OpenStackSwift;
    $.fn.zato.data_table.new_row_func = $.fn.zato.cloud.openstack.swift.data_table.new_row;
    $.fn.zato.data_table.parse();
    $.fn.zato.data_table.setup_forms(
        ['name', 'auth_url', 'auth_version', 'key', 'retries', 'starting_backoff', 'max_backoff', 'pool_size']);
})

$.fn.zato.cloud.openstack.swift.create = function() {
    $.fn.zato.data_table._create_edit('create', 'Create a new OpenStack Swift connection', null);
}

$.fn.zato.cloud.openstack.swift.edit = function(id) {
    $.fn.zato.data_table._create_edit('edit', 'Update the OpenStack Swift connection', id);
}

$.fn.zato.cloud.openstack.swift.data_table.new_row = function(item, data, include_tr) {
    var row = '';

    if(include_tr) {
        row += String.format("<tr id='tr_{0}' class='updated'>", item.id);
    }

    var is_active = item.is_active == true;
    var is_snet = item.is_snet == true;
    var should_validate_cert = item.should_validate_cert == true;
    var should_retr_ratelimit = item.should_retr_ratelimit == true;
    var needs_tls_compr = item.needs_tls_compr == true;

    row += "<td class='numbering'>&nbsp;</td>";
    row += "<td class='impexp'><input type='checkbox' /></td>";
    row += String.format('<td>{0}</td>', item.name);
    row += String.format('<td>{0}</td>', is_active ? 'Yes' : 'No');
    row += String.format('<td>{0}</td>', item.auth_version);
    row += String.format('<td>{0}</td>', item.auth_url);
    row += String.format('<td>{0}</td>', item.user);
    row += String.format('<td>{0}</td>', item.tenant_name);
    row += String.format('<td>{0}</td>', item.pool_size);

    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.cloud.openstack.swift.edit('{0}')\">Edit</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.cloud.openstack.swift.delete_({0});'>Delete</a>", item.id));

    row += String.format("<td class='ignore item_id_{0}'>{0}</td>", item.id);
    row += String.format("<td class='ignore'>{0}</td>", is_active);
    row += String.format("<td class='ignore'>{0}</td>", item.key);
    row += String.format("<td class='ignore'>{0}</td>", item.retries);
    row += String.format("<td class='ignore'>{0}</td>", is_snet);
    row += String.format("<td class='ignore'>{0}</td>", item.starting_backoff);
    row += String.format("<td class='ignore'>{0}</td>", item.max_backoff);
    row += String.format("<td class='ignore'>{0}</td>", should_validate_cert);
    row += String.format("<td class='ignore'>{0}</td>", item.cacert);
    row += String.format("<td class='ignore'>{0}</td>", should_retr_ratelimit);
    row += String.format("<td class='ignore'>{0}</td>", needs_tls_compr);
    row += String.format("<td class='ignore'>{0}</td>", item.custom_options);

    if(include_tr) {
        row += '</tr>';
    }

    return row;
}

$.fn.zato.cloud.openstack.swift.delete_ = function(id) {
    $.fn.zato.data_table.delete_(id, 'td.item_id_',
        'OpenStack Swift connection [{0}] deleted',
        'Are you sure you want to delete the OpenStack Swift connection [{0}]?',
        true);
}

File: code/zato-server/src/zato/server/service/internal/cloud/openstack/swift.py
# -*- coding: utf-8 -*-

"""
Copyright (C) 2019, Zato Source s.r.o. https://zato.io

Licensed under LGPLv3, see LICENSE.txt for terms and conditions.
"""

from __future__ import absolute_import, division, print_function, unicode_literals

# stdlib
from contextlib import closing
from traceback import format_exc

# Zato
from zato.common.broker_message import CLOUD
from zato.common.odb.model import OpenStackSwift
from zato.common.odb.query import cloud_openstack_swift_list
from zato.server.service.internal import AdminService, AdminSIO, GetListAdminSIO

class GetList(AdminService):
    """ Returns a list of OpenStack Swift connections.
    """
    _filter_by = OpenStackSwift.name,

    class SimpleIO(GetListAdminSIO):
        request_elem = 'zato_cloud_openstack_swift_get_list_request'
        response_elem = 'zato_cloud_openstack_swift_get_list_response'
        input_required = ('cluster_id',)
        output_required = (
            'id', 'name', 'is_active', 'auth_url', 'retries', 'starting_backoff', 'max_backoff', 'auth_version', 'key', 'pool_size')
        output_optional = (
            'user', 'is_snet', 'tenant_name', 'should_validate_cert', 'cacert', 'should_retr_ratelimit', 'needs_tls_compr',
            'custom_options')

    def get_data(self, session):
        return self._search(cloud_openstack_swift_list, session, self.request.input.cluster_id, False)

    def handle(self):
        with closing(self.odb.session()) as session:
            self.response.payload[:] = self.get_data(session)

class Create(AdminService):
    """ Creates a new OpenStack Swift connection.
    """
    class SimpleIO(AdminSIO):
        request_elem = 'zato_cloud_openstack_swift_create_request'
        response_elem = 'zato_cloud_openstack_swift_create_response'
        input_required = ('cluster_id', 'name', 'is_active', 'auth_url', 'retries', 'starting_backoff', 'max_backoff',
            'auth_version', 'key', 'should_validate_cert', 'should_retr_ratelimit', 'needs_tls_compr', 'is_snet')
        input_optional = ('user', 'tenant_name', 'cacert', 'custom_options', 'pool_size')
        output_required = ('id', 'name')

    def handle(self):
        input = self.request.input
        with closing(self.odb.session()) as session:
            existing_one = session.query(OpenStackSwift.id).\
                filter(OpenStackSwift.cluster_id==input.cluster_id).\
                filter(OpenStackSwift.name==input.name).\
                first()

            if existing_one:
                raise Exception('An OpenStack Swift connection [{0}] already exists on this cluster'.format(input.name))

            try:
                item = OpenStackSwift()
                for name in self.SimpleIO.input_required + self.SimpleIO.input_optional:
                    setattr(item, name, self.request.input.get(name))

                session.add(item)
                session.commit()

                input.action = CLOUD.OPENSTACK_SWIFT_CREATE_EDIT.value
                self.broker_client.publish(input)

                self.response.payload.id = item.id
                self.response.payload.name = item.name

            except Exception:
                self.logger.error('OpenStack Swift connection could not created, e:`{}`', format_exc())
                session.rollback()

                raise

class Edit(AdminService):
    """ Updates an OpenStack Swift connection.
    """
    class SimpleIO(Create.SimpleIO):
        request_elem = 'zato_cloud_openstack_swift_edit_request'
        response_elem = 'zato_cloud_openstack_swift_edit_response'
        input_required = ('id',) + Create.SimpleIO.input_required

    def handle(self):
        input = self.request.input
        with closing(self.odb.session()) as session:
            existing_one = session.query(OpenStackSwift.id).\
                filter(OpenStackSwift.cluster_id==input.cluster_id).\
                filter(OpenStackSwift.name==input.name).\
                filter(OpenStackSwift.id!=input.id).\
                first()

            if existing_one:
                raise Exception('An OpenStack Swift connection [{0}] already exists on this cluster'.format(input.name))

            try:
                item = session.query(OpenStackSwift).filter_by(id=input.id).one()
                old_name = item.name

                for name in self.SimpleIO.input_required + self.SimpleIO.input_optional:
                    setattr(item, name, self.request.input.get(name))

                session.add(item)
                session.commit()

                input.action = CLOUD.OPENSTACK_SWIFT_CREATE_EDIT.value
                input.old_name = old_name
                self.broker_client.publish(input)

                self.response.payload.id = item.id
                self.response.payload.name = item.name

            except Exception:
                self.logger.error('OpenStack Swift connection could not updated, e:`{}`', format_exc())
                session.rollback()

                raise

class Delete(AdminService):
    """ Deletes an OpenStack Swift connection.
    """
    class SimpleIO(AdminSIO):
        request_elem = 'zato_cloud_openstack_swift_delete_request'
        response_elem = 'zato_cloud_openstack_swift_delete_response'
        input_required = ('id',)

    def handle(self):
        with closing(self.odb.session()) as session:
            try:
                item = session.query(OpenStackSwift).\
                    filter(OpenStackSwift.id==self.request.input.id).\
                    one()

                session.delete(item)
                session.commit()

                msg = {'action': CLOUD.OPENSTACK_SWIFT_DELETE.value, 'name': item.name, 'id':item.id}
                self.broker_client.publish(msg)

            except Exception:
                session.rollback()
                self.logger.error('OpenStack Swift connection could not deleted, e:`{}`', format_exc())

                raise
